rfcs/0007-synchronous-commit.md

```diff
@@ -0,0 +1,162 @@
+
+RocksDB allows mixing writes with different sync settings. For example, if transaction A commits with `sync = false` and transaction B starts afterwards, transaction A's writes will be visible to readers in transaction B. When transaction B commits with `sync = true`, both transactions' writes are persisted. This ordering guarantee means that when a `sync = true` write commits, all previous writes are guaranteed to be persisted as well.
+
+Another important consideration is WAL write is possible to be failures. RocksDB handles these by retrying writes until determining whether the failure is temporary or permanent. In cases of permanent failure (like a full or corrupted disk), RocksDB marks the database state as fatal, rolls back the transaction like nothing happened, and switches the database instance to read-only mode.
+
+## Synchronous Commit in a summary 
+
+Based on the PostgreSQL and RocksDB references above, we can summarize the key semantics of Synchronous Commit:
+
+1. A write is only considered committed once the WAL has been persisted to storage. Until then, the data remains invisible to readers.
+2. If there is a permanent failure while persisting the WAL during a Synchronous Commit, the transaction rolls back completely. The database instance enters a fatal state and switches to read-only mode.
+3. Multiple levels of Synchronous Commit can be supported, allowing users to balance performance and durability requirements.
+4. Synchronous and Unsynchronous Commits can be interleaved in different transactions. A transaction using Synchronous Commit can read writes from transactions that used Unsynchronous Commit. When a Synchronous Commit persists, it also persists any previous Unsynchronous Commit writes in the WAL.
+
+## Current Design in SlateDB
+
+This section is based on @criccomini 's comment in <https://github.com/slatedb/slatedb/pull/260#issuecomment-2576502212>.
+
+SlateDB currently does not provide an explicit notion of Synchronous Commit. But it does provide a `DurabilityLevel` enum to control the durability guarantees on both read and write operations.
+
+The `DurabilityLevel` enum is defined as follows:
+
+```rust
+enum DurabilityLevel {
+    Memory,
+    Local, // not implemented yet
+    Remote,
+}
+```
+
+And the `WriteOptions` struct contains a `await_durability: DurabilityLevel` option to control the waiting behavior for durability. If `await_durability` is set to `DurabilityLevel::Remote`, the write will wait for the WAL to be written into S3 before returning.
+
+Please note that the commit semantic is a bit different from other systems' Synchronous Commit. No matter what `DurabilityLevel` is set in the write operation, this write is considered visible to the readers with `DurabilityLevel::Memory` immediately after the write is appended to the WAL, not nessarily flushed to storage.
```

Comment:
I would say `await_durability` is not even a "commit semantic". It is only saying something about durability: "I will return when the data has hit memory/object storage." That's it. I keep hammering this because I think it's important to recognize that durability and commit semantics are two separate things.

Reply:
make sense, "commit semantic" often contains durability guarantee at some level, but durability guarantee does not necessarily means it's a "commit" 😲